### PR TITLE
[BugFix] Forbid list partition for backup/restore feature

### LINKED ISSUE #27324: [auto create list partition] backup restore list partition table，partition missing
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
```
create database auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b;
use auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b
CREATE TABLE `duplicate_table_without_null` ( `k1`  date not null, `k2`  datetime not null,`k3`  char(20) not null, `k4`  varchar(20) not null, `k5`  string not null, `k6`  boolean not null, `k7`  tinyint not null, `k8`  smallint not null, `k9`  int not null, `k10`  bigint not null, `k11` largeint not null, `k12` decimal(27,9) not null, `k13` int not null) DUPLICATE KEY(`k1`, `k2`, `k3`, `k4`, `k5`) PARTITION BY (k1) DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) PROPERTIES (  "replication_num" = "3", "storage_format" = "v2");

insert into duplicate_table_without_null values('2020-01-01', '2020-01-01 12:30:30', 'asfgrgte', 'wergergqer', 'stringaaa', 0, 11, 111, 1111, 11111, 111111, 11.11, 1), ('2020-01-02', '2020-01-02 12:30:30', 'asfgrgtf', 'wergergqes', 'stringbbb',1, 22, 222, 2222, 22222, 222222, 22.22, 2);
show partitions from duplicate_table_without_null order by PartitionName;


CREATE REPOSITORY `backup_repository_7733ec6c_2447_11ee_9db5_00163e0e550b` with broker `hdfs_broker` on location (http://xxx/user/starrocks_test_tmp/backup_repository_7733ec6c_2447_11ee_9db5_00163e0e550b)" properties ( "username"= "sr","password" = "");
BACKUP SNAPSHOT auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b.test_backup to backup_repository_7733ec6c_2447_11ee_9db5_00163e0e550b PROPERTIES ('timeout' = '600');

drop database auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b;
create database auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b;


show snapshot on backup_repository_7733ec6c_2447_11ee_9db5_00163e0e550b;
RESTORE SNAPSHOT auto_set_bucket_768123ac_2447_11ee_9db5_00163e0e550b.`test_backup` from `backup_repository_7733ec6c_2447_11ee_9db5_00163e0e550b` PROPERTIES( "backup_timestamp"="2023-07-17-10-13-05-671", "replication_num" = "3", "timeout" = "600");

show partitions from duplicate_table_without_null order by PartitionName;
```

image

e17f4cfa-a38d-4f55-8587-ef7276127c1d

### Expected behavior (Required)
partition exists
### Real behavior (Required)
partition missing

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
mysql> select current_version();
+--------------------+
| current_version()  |
+--------------------+
| branch-3.1-e3010ef |
+--------------------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by Astralidea**
Fixes #27324

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java|6|7|85.71%|[316]|

**Review by ABingHuang — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java [hunk 1]
```diff
@@ -306,12 +308,19 @@ public static void analyzeTableRef(TableRef tableRef, String dbName, Database db
...
+        OlapTable olapTbl = null;
+        if (tbl instanceof OlapTable) {
+            olapTbl = (OlapTable) tbl;
+            PartitionInfo partitionInfo = olapTbl.getPartitionInfo();
+            if (partitionInfo instanceof ListPartitionInfo) {
+                throw new SemanticException("List partition table does not support backup/restore job");
+            }
+        }
+
+        if (olapTbl != null && partitionNames != null) {
```
Comment (srlch) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java:320]: If olapTbl != null, !tbl.isOlapOrCloudNativeTable() is always false, it never hit the exception here. Maybe we should remove the olapTbl != null here?

**Pull request merged**